### PR TITLE
Add reconnecting udp connection type to jaeger exporter

### PR DESCRIPTION
**Pull request by terev**
This pr ports this change https://github.com/jaegertracing/jaeger-client-go/pull/520 to the open-telemetry Jaeger exporter. The goal of the change is to detect changes to an agent's host dns record and adopt the new address as the destination address.

## Short description of the changes
- This pr adds a new type that decorates a UDPConn
- This type attempts to resolve a new address for the given host, if a new address is resolved, and the dial succeeds the currently held connection is atomically swapped with a new one

### CODE REVIEW

File: exporters/trace/jaeger/agent.go [hunk 1]
```diff
@@ -32,41 +34,76 @@ type agentClientUDP struct {
...
-func newAgentClientUDP(hostPort string, maxPacketSize int) (*agentClientUDP, error) {
-	if maxPacketSize == 0 {
-		maxPacketSize = udpPacketMaxLength
+func newAgentClientUDP(params agentClientUDPParams) (*agentClientUDP, error) {
+	// validate hostport
+	if _, _, err := net.SplitHostPort(params.HostPort); err != nil {
+		return nil, err
+	}
+
+	if params.MaxPacketSize == 0 {
```
Comment (Aneurysm9) [inline exporters/trace/jaeger/agent.go:64]:
```suggestion
	if params.MaxPacketSize <= 0 {
```

File: exporters/trace/jaeger/agent.go [hunk 2]
```diff
@@ -32,41 +34,76 @@ type agentClientUDP struct {
...
+		params.MaxPacketSize = udpPacketMaxLength
 	}
 
-	thriftBuffer := thrift.NewTMemoryBufferLen(maxPacketSize)
+	if !params.DisableAttemptReconnecting && params.AttemptReconnectInterval == 0 {
```
Comment (Aneurysm9) [inline exporters/trace/jaeger/agent.go:68]:
```suggestion
	if !params.DisableAttemptReconnecting && params.AttemptReconnectInterval <= 0 {
```

File: exporters/trace/jaeger/agent.go [hunk 3]
```diff
@@ -32,41 +34,76 @@ type agentClientUDP struct {
...
+	Write([]byte) (int, error)
+	SetWriteBuffer(int) error
+	Close() error
+}
+
+type agentClientUDPParams struct {
+	HostPort                   string
+	MaxPacketSize              int
+	Logger                     *log.Logger
+	DisableAttemptReconnecting bool
```
Comment (Aneurysm9) [inline exporters/trace/jaeger/agent.go:53]:
```suggestion
	AttemptReconnecting bool
```

Stating this in the affirmative and negating it where necessary maintains consistency with the `AttemptReconnectInterval` field and avoids double negation that can require excess cognitive load when interpreting statements referring to this field.

File: exporters/trace/jaeger/reconnecting_udp_client.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+func (c *reconnectingUDPConn) Write(b []byte) (int, error) {
+	var bytesWritten int
+	var err error
+
+	c.connMtx.RLock()
+	if c.conn == nil {
+		// if connection is not initialized indicate this with err in order to hook into retry logic
+		err = fmt.Errorf("UDP connection not yet initialized, an address has not been resolved")
+	} else {
+		bytesWritten, err = c.conn.Write(b)
```
Comment (jmacd) [inline exporters/trace/jaeger/reconnecting_udp_client.go:146]: I would prefer to see the `Write()` happen without holding the lock. Can we release the lock after reading `c.conn`?

Comment (terev) [reply exporters/trace/jaeger/reconnecting_udp_client.go:146]: It's an rlock so concurrent writes won't block . The lock is necessary in order to ensure all rlocks are drained by a write lock before swapping the connection pointer

Comment (jmacd) [reply exporters/trace/jaeger/reconnecting_udp_client.go:146]: There are well-known problems with RWLocks though. The concurrent calls to write won't block, but when many `Write()` calls are happening, calls to `attemptDialNewAddr` and `Close()` can block a very long time.

Swapping the connection pointer doesn't need to be synchronized here, only closing the old socket does. But when you're closing the old socket that means the socket is already in a questionable state, right? (I believe this means you're synchronizing on a write that you expect to fail.)

Comment (jmacd) [reply exporters/trace/jaeger/reconnecting_udp_client.go:146]: (I do not want to block this PR, but it's suspicious to hold a read lock while making a system call, especially when the lock doesn't look necessary--but I may be wrong about that.)

Comment (terev) [reply exporters/trace/jaeger/reconnecting_udp_client.go]: @jmacd This is a very good point, thanks i've learned something here 😄  . It looks like the UDPConn implementation does synchronization itself with a rwlock so calling Close should be safe regardless of whether there's an ongoing Write. I've made the change to read the conn then make the Write call

File: exporters/trace/jaeger/uploader.go [hunk 5]
```diff
@@ -36,13 +38,24 @@ type EndpointOption func() (batchUploader, error)
 
 // WithAgentEndpoint instructs exporter to send spans to jaeger-agent at this address.
 // For example, localhost:6831.
-func WithAgentEndpoint(agentEndpoint string) EndpointOption {
+func WithAgentEndpoint(agentEndpoint string, options ...AgentEndpointOption) EndpointOption {
```
Comment (MrAlias) [inline exporters/trace/jaeger/uploader.go:41]: This will need to be included in the cleanup of #536

**Pull request merged**